cmd/nitro/init.go

```diff
@@ -606,132 +606,112 @@ func rebuildLocalWasm(ctx context.Context, config *gethexec.Config, l2BlockChain
-				if err != nil {
-					return executionDB, nil, fmt.Errorf("error pruning: %w", err)
-				}
-				l2BlockChain, err := gethexec.GetBlockChain(executionDB, cacheConfig, chainConfig, tracer, &config.Execution.TxIndexer)
-				if err != nil {
-					return executionDB, nil, err
-				}
-				err = validateBlockChain(l2BlockChain, chainConfig)
-				if err != nil {
-					return executionDB, l2BlockChain, err
-				}
-				if config.Init.RecreateMissingStateFrom > 0 {
-					err = staterecovery.RecreateMissingStates(executionDB, l2BlockChain, cacheConfig, config.Init.RecreateMissingStateFrom)
-					if err != nil {
-						return executionDB, l2BlockChain, fmt.Errorf("failed to recreate missing states: %w", err)
-					}
-				}
-				return rebuildLocalWasm(ctx, &config.Execution, l2BlockChain, executionDB, wasmDB, config.Init.RebuildLocalWasm)
-			}
-			readOnlyDB.Close()
-		} else if !dbutil.IsNotExistError(err) {
-			// we only want to continue if the database does not exist
-			return nil, nil, fmt.Errorf("Failed to open database: %w", err)
-		}
+func openInitializeExecutionDB(ctx context.Context, stack *node.Node, config *NodeConfig, chainId *big.Int, cacheConfig *core.BlockChainConfig, tracer *tracing.Hooks, persistentConfig *conf.PersistentConfig, l1Client *ethclient.Client, rollupAddrs chaininfo.RollupAddresses) (ethdb.Database, *core.BlockChain, error) {
+	executionDB, wasmDB, l2BlockChain, chainConfig, err := openExistingExecutionDB(stack, config, chainId, cacheConfig, tracer, persistentConfig)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	// Check if database was misplaced in parent dir
-	const errorFmt = "database was not found in %s, but it was found in %s (have you placed the database in the wrong directory?)"
-	parentDir := filepath.Dir(stack.InstanceDir())
-	if dirExists(path.Join(parentDir, "l2chaindata")) {
-		return nil, nil, fmt.Errorf(errorFmt, stack.InstanceDir(), parentDir)
-	}
-	grandParentDir := filepath.Dir(parentDir)
-	if dirExists(path.Join(grandParentDir, "l2chaindata")) {
-		return nil, nil, fmt.Errorf(errorFmt, stack.InstanceDir(), grandParentDir)
-	}
+	if executionDB == nil {
+		err := checkAndDownloadDB(ctx, stack, config)
+		if err != nil {
+			return nil, nil, err
+		}
 
-	if err := checkEmptyDatabaseDir(stack.InstanceDir(), config.Init.Force); err != nil {
-		return nil, nil, err
-	}
+		executionDB, wasmDB, err = openDownloadedExecutionDB(stack, config, cacheConfig, persistentConfig)
+		if err != nil {
+			return nil, nil, err
+		}
 
-	if err := setLatestSnapshotUrl(ctx, &config.Init, config.Chain.Name); err != nil {
-		return nil, nil, err
-	}
+		var initDataReader *statetransfer.InitDataReader
+		var genesisArbOSInit *params.ArbOSInit
 
-	initFile, cleanUpTmp, err := initializeAndDownloadInit(ctx, &config.Init, stack)
-	defer cleanUpTmp()
-	if err != nil {
-		return nil, nil, err
-	}
+		initDataReader, chainConfig, genesisArbOSInit, err = getInit(config)
```

Comment:
This func computes its output based on config.Init.
To make that clear, getInit could receive config.Init instead of config as argument.
And also be renamed to something like getInitFromNodeConfig.
WDYT?